sqs/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/SqsSink.scala

```diff
@@ -0,0 +1,335 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.collectors.scalastream.sinks
+
+import com.amazonaws.{AmazonClientException, AmazonWebServiceRequest, ClientConfiguration}
+import com.amazonaws.auth.{
+  AWSCredentialsProvider,
+  AWSStaticCredentialsProvider,
+  BasicAWSCredentials,
+  DefaultAWSCredentialsProviderChain,
+  EnvironmentVariableCredentialsProvider,
+  InstanceProfileCredentialsProvider
+}
+import com.amazonaws.retry.{PredefinedBackoffStrategies, RetryPolicy}
+import com.amazonaws.retry.RetryPolicy.RetryCondition
+import com.amazonaws.services.sqs.{AmazonSQS, AmazonSQSClientBuilder}
+import cats.syntax.either._
+import com.amazonaws.services.sqs.model.{
+  QueueDoesNotExistException,
+  SendMessageBatchRequest,
+  SendMessageBatchRequestEntry
+}
+import com.snowplowanalytics.snowplow.collectors.scalastream.model._
+import org.slf4j.LoggerFactory
+
+import java.nio.ByteBuffer
+import java.util.UUID
+import java.util.concurrent.ScheduledExecutorService
+import scala.concurrent.{ExecutionContextExecutorService, Future}
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.MILLISECONDS
+import scala.util.{Failure, Success}
+
+/**
+  *  SQS sink for the Scala Stream Collector
+  */
+class SqsSink private (
+  client: AmazonSQS,
+  sqsConfig: Sqs,
+  bufferConfig: BufferConfig,
+  queueName: String,
+  executorService: ScheduledExecutorService
+) extends Sink {
+  // Records must not exceed 256K when writing to SQS.
+  // Since messages can be base64-encoded, they must meet that limit after encoding.
+  override val MaxBytes = 192000 // 256000 / 4 * 3
+
+  val ByteThreshold: Long   = bufferConfig.byteLimit
+  val RecordThreshold: Long = bufferConfig.recordLimit
+  val TimeThreshold: Long   = bufferConfig.timeLimit
+
+  private val maxBackoff      = sqsConfig.backoffPolicy.maxBackoff
+  private val minBackoff      = sqsConfig.backoffPolicy.minBackoff
+  private val randomGenerator = new java.util.Random()
+
+  implicit lazy val ec: ExecutionContextExecutorService =
+    concurrent.ExecutionContext.fromExecutorService(executorService)
+
+  def storeRawEvents(events: List[Array[Byte]], key: String): List[Array[Byte]] = {
+    events.foreach(e => EventBuffer.store(e))
+    Nil
+  }
+
+  object EventBuffer {
+    type Event = ByteBuffer
+
+    private var storedEvents              = List.empty[Event]
+    private var byteCount                 = 0L
+    @volatile private var lastFlushedTime = 0L
+
+    def store(event: Array[Byte]): Unit = {
+      val eventBytes = ByteBuffer.wrap(event)
+      val eventSize  = eventBytes.capacity
+      if (eventSize >= MaxBytes) {
+        log.error(
+          s"Message of size $eventSize bytes is too large. It must be less than $MaxBytes bytes."
```

Comment:
We do. The CollectorService is in charge of that. By the time this function is called, we've already created the bad rows (if any). So for the good sink, we should never see this message. For the bad sink, it might appear in theory, if the bad row we are trying to write is too big. If that is the case, we log this error, to alert us that something is wrong.

Reply:
> For the bad sink, it might appear in theory, if the bad row we are trying to write is too big

In this case we can resize the size of the payload inside the bad row, so that it's small enough and it can be emitted

> we log this error, to alert us

In this case an input event doesn't have an associated output (either good or bad)

Reply:
> In this case we can resize the size of the payload

Yes, actually, that is exactly what we do: https://github.com/snowplow/stream-collector/blob/master/core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/utils/SplitBatch.scala#L143

Reply:
Also for `generic_error` ? I'm asking but I don't know if bad rows of this type can be very big

Reply:
No, we don't have that for `generic_error` but the only way it can get too big is if the query string is massive and I think browser and server limits will kick in way before it becomes an issue.